src/Stratis.Bitcoin/Consensus/ProvenHeadersConsensusManagerBehavior.cs

```diff
@@ -55,9 +66,57 @@ protected override async Task OnMessageReceivedAsync(INetworkPeer peer, Incoming
                 case GetProvenHeadersPayload getHeaders:
                     await this.ProcessGetHeadersAsync(peer, getHeaders).ConfigureAwait(false);
                     break;
+                case VersionPayload version:
+                    await this.ProcessVersionAsync(peer, version).ConfigureAwait(false);
+                    break;
             }
         }
 
+        /// <summary>
+        /// Processes "version" message received from the peer.
+        /// Ensures we leave some available connection slots to nodes that are enabled to serve Proven Headers, if the number of minimum PH enabled nodes connected isn't met.
+        /// </summary>
+        /// <param name="peer">Peer from which the message was received.</param>
+        /// <param name="version">Payload of "version" message to process.</param>
+        protected Task ProcessVersionAsync(INetworkPeer peer, VersionPayload version)
+        {
+            // We enforce having free slots for PH enabled nodes only when we are not in IBD.
+            if (!this.initialBlockDownloadState.IsInitialBlockDownload())
+            {
+                PeerConnectorDiscovery connector = this.connectionManager.PeerConnectors.OfType<PeerConnectorDiscovery>().FirstOrDefault();
+                // If PeerConnectorDiscovery is not found means we are using other ways to connect peers (like -connect) and thus we don't enforce the rule.
+                if (connector != null)
+                {
+                    int slotsReservedForProvenHeaderEnabledPeers = (int)Math.Round(connector.MaxOutboundConnections * ProvenHeaderPeersReservedSlotsThreshold, MidpointRounding.ToEven);
+                    int maxLegacyPeersAllowed = connector.MaxOutboundConnections - slotsReservedForProvenHeaderEnabledPeers;
+                    int legacyPeersConnectedCount = connector.ConnectorPeers
+                        .Where(p => p.PeerVersion.Version < NBitcoin.Protocol.ProtocolVersion.PROVEN_HEADER_VERSION)
+                        .Count();
+
+                    bool hasToReserveSlots = legacyPeersConnectedCount >= maxLegacyPeersAllowed;
+                    if (hasToReserveSlots)
+                    {
+                        // If we were previously in IBD state, we should remove legacy peers to reserve slots for PH enabled peers.
+                        int legacyPeersToDisconnect = legacyPeersConnectedCount - maxLegacyPeersAllowed;
+                        if (legacyPeersToDisconnect > 0)
+                        {
+                            var peersToDisconnect = connector.ConnectorPeers.OrderBy(p => p.PeerVersion.StartHeight).Take(legacyPeersToDisconnect).ToList();
```

Comment:
`p.PeerVersion.StartHeight` is not a good indicator, you need to find the correct `ProvenHeadersConsensusManagerBehavior` for that peer and look at its tip also you are not reading the correct linq query you need to filter out `connector.ConnectorPeers` bellow `PROVEN_HEADER_VERSION` also here right? if not you might disconnect PH enabled nodes.

But I think ther eis also another way to check if a peer is a PH peer

Reply:
Also we shouldn't care about start height as long as policy is activated after we only can sync with PH and those nodes can't do it no matter the height

Reply:
I ended up doing this helper method 
```
private IEnumerable<INetworkPeer> GetConnectedLegacyPeersSortedByTip(NetworkPeerCollection connectedPeers)
{
    return from peer in connectedPeers.ToList() // not sure if connectedPeers can change, so i use ToList to get a snapshot
            let isLegacy = peer.PeerVersion.Version < NBitcoin.Protocol.ProtocolVersion.PROVEN_HEADER_VERSION
            let tip = peer.Behavior<ProvenHeadersConsensusManagerBehavior>()?.ExpectedPeerTip?.Height ?? 0
            where isLegacy
            orderby tip
            select peer;
}
```

Reply:
yeah `connectedPeers` can suddenly change so good you took a snapshot.

Ok this linq query is nice but how easy will it be to change it in the future?
When we refactor `Protocol.ProtocolVersion` this will have to change a bit.

But this looks correct